### PR TITLE
Machine config changes

### PR DESCRIPTION
**Pull request by alsrdn**
# Reason for This PR

Fixes several issues with machine_config:
- The following fields on machine-config: `vcpu_count` and `mem_size_mib` are mandatory when configuring via the API but optional when configuring from a JSON file.
- The CPU template and `smt` can be configured on ARM as well when starting from a JSON config. It only returns an error when using the API for config.
- If a machine_config PUT request is issued without the `track_dirty_pages` and the `cpu_template` parameters set, track_dirty_pages would be set to false but the cpu_template would remain unchanged.
- There is no way to reset the cpu_template through PUT or PATCH once it has been set with either of these methods.
- Patching track_dirty_pages was not allowed on aarch64 even though it is a pre-boot request.

## Description of Changes

This patch splits the actual machine configuration in 2 structs: one used for keeping the actual configured values and one used only for updates.
Since Firecracker starts with default values for the machine configuration, all changes (either PUT or PATCH) are unified through an update path.
Having 2 distinct structs allows for validation of parameters early in the deserialization of the structs, and eliminates the need for handling unwraps for the parameters in the VMM.
Since the parameters are validated in the deserializer, we can use the same code for PUT, PATCH and starting from a json file.

Also added more details in the swagger file to better explain the behaviour of PUT.

This change requires a minor version upgrade.

- [ ] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [ ] The issue which led to this PR has a clear conclusion.
- [ ] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

**Review by serban300 — APPROVED** [review]

**Review by georgepisaltu — APPROVED** [review]

**Review by alindima — COMMENTED** [review]
Let's also add a python regression tests for configuring from the JSON config file (making sure that mem_size_mib and vcpu_count are required) and that cpu_template and smt cannot be configured

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,38 @@
...
+
+- Fixed inconsistency that allowed the start of a microVM from a JSON file
+  without specifying the `vcpu_count` and `mem_size_mib` parameters for
+  `machine-config` although they are mandatory when configuring via the API.
+  Now these fields are mandatory when specifying `machine-config` in the JSON
+  file and when using the `PUT` request on `/machine-config`.
+- Fixed inconsistency that allowed a user to specify the `smt` and
+  `cpu_template` parameters in `machine-config` when starting from a JSON file
+  on aarch64 although they are not permitted when using `PUT` or `PATCH` in
+  the API. Now Firecracker will return an error if these parameters are used
```
[the remark this answers is not in the document]
Comment (alsrdn) [reply CHANGELOG.md:26]: Yes, I need to adjust the description in the Changelog.

File: src/api_server/swagger/firecracker.yaml [hunk 2]
```diff
@@ -302,9 +302,12 @@ paths:
       description:
         Updates the Virtual Machine Configuration with the specified input.
         Firecracker starts with default values for vCPU count (=1) and memory size (=128 MiB).
-        With SMT enabled, the vCPU count is restricted to be 1 or an even number,
+        The vCPU count is restricted to the [1, 32] range.
+        With SMT enabled, the vCPU count is required to be either 1 or an even number in the range.
         otherwise there are no restrictions regarding the vCPU count.
         If any of the parameters has an incorrect value, the whole update fails.
+        All parameters that are optional and are not specified are set to their default values
+        (smt = false, track_dirty_pages = false, cpu_template = None).
```
[the remark this answers is not in the document]
Comment (alsrdn) [reply src/api_server/swagger/firecracker.yaml:310]: I've added a default for track_dirty_pages in the swagger definition. Will need to add one for CPU template as well.

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (alsrdn) [reply CHANGELOG.md]: Changed changelog.

File: src/api_server/src/request/machine_configuration.rs [hunk 3]
```diff
@@ -19,75 +19,39 @@ pub(crate) fn parse_put_machine_config(body: &Body) -> Result<ParsedRequest, Err
...
     })?;
 
-    #[cfg(target_arch = "aarch64")]
-    check_unsupported_fields(&vm_config)?;
-
     if vm_config.vcpu_count.is_none()
         && vm_config.mem_size_mib.is_none()
         && vm_config.cpu_template.is_none()
         && vm_config.smt.is_none()
+        && vm_config.track_dirty_pages.is_none()
```
[the remark this answers is not in the document]
Comment (alsrdn) [reply src/api_server/src/request/machine_configuration.rs]: Added an is_empty method.

**Comment by alsrdn** [comment]
> Let's also add a python regression tests for configuring from the JSON config file (making sure that mem_size_mib and vcpu_count are required) and that cpu_template and smt cannot be configured on ARM

I've added some more integration tests for starting with invalid machine config parameters in the json.

File: tests/integration_tests/functional/test_cmd_line_start.py [hunk 4]
```diff
@@ -191,6 +192,67 @@ def test_config_start_no_api(test_microvm_with_api, vm_config_file):
         delay=1)
 
 
+@pytest.mark.parametrize(
+    "vm_config_file",
+    [
+        "framework/vm_config_missing_vcpu_count.json",
+        "framework/vm_config_missing_vcpu_mem_size_mib.json"
```
Comment (alindima) [inline tests/integration_tests/functional/test_cmd_line_start.py:199]: Do you mean:
```suggestion
        "framework/vm_config_missing_mem_size_mib.json"
```
?

File: tests/framework/vm_config_missing_vcpu_mem_size_mib.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    "vcpu_count": 2,
+    "smt": false,
+    "track_dirty_pages": false
+  },
+  "balloon": null,
+  "network-interfaces": [],
+  "vsock": null,
+  "logger": null,
+  "metrics": null,
+  "mmds-config": null
```
Comment (alindima) [inline tests/framework/vm_config_missing_vcpu_mem_size_mib.json:29]: nit: all of those null entries are not mandatory, right? So we may remove them to keep the file more focused
Same for specific configs like the drive cache_type

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,38 @@
...
+- Fixed inconsistency that allowed the start of a microVM from a JSON file
+  without specifying the `vcpu_count` and `mem_size_mib` parameters for
+  `machine-config` although they are mandatory when configuring via the API.
+  Now these fields are mandatory when specifying `machine-config` in the JSON
+  file and when using the `PUT` request on `/machine-config`.
+- Fixed inconsistency that allowed a user to specify the `cpu_template`
+  parameter and set `smt` to `True` in `machine-config` when starting from a
+  JSON file on aarch64 even though they are not permitted when using `PUT` or
+  `PATCH` in the API. Now Firecracker will return an error if `smt` is set to
+  `True` or if `cpu_template` is specified on aarch64.
```
Comment (alindima) [inline CHANGELOG.md:27]: nit: 
```suggestion
  `True` on aarch64 or if `cpu_template` is specified on aarch64.
```

It's repetitive but I think it's more clear

**Review by alindima — COMMENTED** [review]
Looks good

File: tests/framework/vm_config_missing_vcpu_mem_size_mib.json (hunk 5 above)
Comment (alsrdn) [reply tests/framework/vm_config_missing_vcpu_mem_size_mib.json]: ok, I have minimised the json configs.

File: CHANGELOG.md (hunk 6 above)
Comment (alsrdn) [reply CHANGELOG.md]: Ok, rephrased.

**Review by serban300 — APPROVED** [review]

**Review by georgepisaltu — APPROVED** [review]

**Pull request merged**